### PR TITLE
AP_GPS: add support for SBF, novatel and unicore dual antenna GPS

### PR DESCRIPTION
**Pull request by Vkd4U2NtVkZOVg**
I close those two PR: #18172 #18492 , and combine to this one.

### CODE REVIEW

**Comment by Vkd4U2NtVkZOVg** [comment]
rebase to the master.

**Comment by Vkd4U2NtVkZOVg** [comment]
tested in the real flight，both GPS's yaw works fine.

**Comment by timsonemma** [comment]
> Has this been tested with Septentrio Mosaic-H or just AsteRx series hardware from them?

Yes this has been tested with Mosaic-H

**Comment by timsonemma** [comment]
> @timsonemma anything else should we do before merge this pr?

seems good to me!

File: libraries/AP_GPS/AP_GPS_SBF.cpp [hunk 1]
```diff
@@ -112,12 +112,20 @@ AP_GPS_SBF::read(void)
...
                                 break;
                             case Config_State::SSO:
-                                if (asprintf(&config_string, "sso,Stream%d,COM%d,PVTGeodetic+DOP+ReceiverStatus+VelCovGeodetic+BaseVectorGeod,msec100\n",
-                                             (int)GPS_SBF_STREAM_NUMBER,
-                                             (int)gps._com_port[state.instance]) == -1) {
-                                    config_string = nullptr;
-                                }
-                                break;
+                                if (get_type() == AP_GPS::GPS_TYPE_SBF) {
+                                    if (asprintf(&config_string, "sso,Stream%d,COM%d,PVTGeodetic+DOP+ReceiverStatus+VelCovGeodetic+BaseVectorGeod,msec100\n",
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_SBF.cpp:116]: The alternate form of this that would use less flash is to pass the extra arguments as a %s at the end of the string like this: `"sso,Stream%d,COM%d,PVTGeodetic+DOP+ReceiverStatus+VelCovGeodetic+BaseVectorGeod%s,msec100\n"` then pass a string that is either empty "" or contains "AttEuler+AttCovEuler" depending on which type it is (which can be done as a separate switch before the asprintf call.  This is both quite readable, reduces flash cost, and most importantly it makes it easier to add (or remove) more required terms from all the driver cases, without having to duplicate it into every string manually.

File: libraries/AP_GPS/AP_GPS_NOVA.cpp [hunk 2]
```diff
@@ -257,6 +278,42 @@ AP_GPS_NOVA::process_message(void)
...
+                case 33: // iono float
+                case 34: // narrow float
+                case 48: // l1 int
+                case 50: // narrow int
+                {
+                    state.have_gps_yaw = true;
+                    state.gps_yaw_configured = true;
+                    state.have_gps_yaw_accuracy = true;
+                    state.gps_yaw = (float)(headingu.heading);
+                    state.gps_yaw_accuracy = (float)(headingu.hdgstddev);
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_NOVA.cpp:304]: Why do these need casts to float? The struct member is float as far as I can see.

File: libraries/AP_GPS/AP_GPS_NOVA.cpp [hunk 3]
```diff
@@ -257,6 +278,42 @@ AP_GPS_NOVA::process_message(void)
         _new_position = true;
     }
 
+    // 971 for unicore 2042 for novatel
+    if(messageid == 971 || messageid == 2042)
+    {
+        const nova_heading &headingu = nova_msg.data.headingu;
+        if (headingu.solstat == 0) {// have a solution
+            switch (headingu.postype)
+            {
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_NOVA.cpp:287]: Unneeded extra line

File: libraries/AP_GPS/AP_GPS_NOVA.cpp [hunk 4]
```diff
@@ -257,6 +278,42 @@ AP_GPS_NOVA::process_message(void)
         _new_position = true;
     }
 
+    // 971 for unicore 2042 for novatel
+    if(messageid == 971 || messageid == 2042)
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_NOVA.cpp:282]: I think you need to actually only enter this if it's unicore and 971 or, novatel and 2042.

File: libraries/AP_GPS/AP_GPS_NOVA.cpp [hunk 5]
```diff
@@ -67,14 +67,35 @@ AP_GPS_NOVA::read(void)
...
+                    port->write((const uint8_t *)unicore_heading_init_blob, strlen(unicore_heading_init_blob));
+                    break;
+                default:
+                    break;
+            }
+            _init_blob_index++;
+        }
+    }
+
+    if (now - state.gps_yaw_time_ms > 500) {
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_NOVA.cpp:97]: Too long a timeout period.

File: libraries/AP_GPS/AP_GPS_NOVA.cpp [hunk 6]
```diff
@@ -67,14 +67,35 @@ AP_GPS_NOVA::read(void)
...
 
         if (now > _init_blob_time) {
             port->write((const uint8_t*)init_str, strlen(init_str));
             _init_blob_time = now + 200;
             _init_blob_index++;
         }
+    } else if (_init_blob_index == ARRAY_SIZE(_initialisation_blob)) {
+        if (now > _init_blob_time) {
+            switch (get_type())
+            {
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_NOVA.cpp:81]: This should be on the same line as above.

**Review by WickedShell — CHANGES_REQUESTED** [review]
Please don't combine PR's like this. The pair of combined PR's are actually very different, and merging them like this makes it take longer to actually get in, as well as makes it harder to track the history.

Has the modified novatel driver been run against novatel hardware after the changes?

You also need to add the new options to the GPS_TYPE parameter descriptions in AP_GPS.cpp.

I may come back and try and split the SBF PR back out, later with changes, unless you beat me to it.

File: libraries/AP_GPS/AP_GPS_NOVA.cpp [hunk 7]
```diff
@@ -67,14 +67,35 @@ AP_GPS_NOVA::read(void)
...
+                case AP_GPS::GPS_TYPE_NOVA:
+                    break;
+                case AP_GPS::GPS_TYPE_NOVA_DUAL:
+                    port->write((const uint8_t *)nova_heading_init_blob, strlen(nova_heading_init_blob));
```
[the remark this answers is not in the document]
Comment (Vkd4U2NtVkZOVg) [reply libraries/AP_GPS/AP_GPS_NOVA.cpp:85]: I think I can handle the split job. I'm gonna remove the novatel driver's modification and focus to the sbf driver as you suggested.

File: libraries/AP_GPS/AP_GPS_SBF.cpp [hunk 8]
```diff
@@ -497,6 +508,33 @@ AP_GPS_SBF::process_message(void)
...
+        }
+        break;
+    }
+
+    case AttCovEuler:
+    {
+        const msg5939 &temp = sbf_msg.data.msg5939u;
+        check_new_itow(temp.TOW, sbf_msg.length);
+        if (!(temp.Aux1Error || temp.Aux2Error)) {
+            state.gps_yaw_accuracy = (float)(sqrt(temp.Cov_HeadHead));
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_SBF.cpp:532]: You also need to check for the DNU value before trying this.

File: libraries/AP_GPS/AP_GPS_SBF.cpp [hunk 9]
```diff
@@ -497,6 +508,33 @@ AP_GPS_SBF::process_message(void)
...
+
+    case AttEuler:
+    {
+        const msg5938 &temp = sbf_msg.data.msg5938u;
+        check_new_itow(temp.TOW, sbf_msg.length);
+
+        if (!(temp.Aux1Error || temp.Aux2Error) && temp.Mode && (temp.Heading > -200000)) {
+            state.have_gps_yaw = true;
+            state.gps_yaw_configured = true;
+            state.gps_yaw = (float)(temp.Heading);
```
Comment (WickedShell) [inline libraries/AP_GPS/AP_GPS_SBF.cpp:520]: Check for DNU before copying it in.

**Pull request closed without merging**